### PR TITLE
toaplan2.cpp: Split class per hardware feature

### PR DESCRIPTION
**Pull request by cam900**

### CODE REVIEW

File: src/mame/toaplan/toaplan2.h [hunk 1]
```diff
@@ -172,81 +238,64 @@ class toaplan2_state : public driver_device
...
+		, m_rtc(*this, "rtc")
+		, m_hopper(*this, "hopper")
+	{ }
+
+	void othldrby(machine_config &config);
+	void pwrkick(machine_config &config);
+
+private:
+	optional_device<upd4992_device> m_rtc;
+	optional_device<ticket_dispenser_device> m_hopper;
```
Comment (cuavas) [inline src/mame/toaplan/toaplan2.h:292]: At least the RTC could be made required.

File: src/mame/toaplan/toaplan2.cpp [hunk 2]
```diff
@@ -416,10 +416,15 @@ constexpr unsigned toaplan2_state::T2PALETTE_LENGTH;
...
 {
 	save_item(NAME(m_old_p1_paddle_h));
 	save_item(NAME(m_old_p2_paddle_h));
+}
+
+
+void truxton2_state::machine_start()
+{
 	save_item(NAME(m_z80_busreq));
 }
```
Comment (cuavas) [inline src/mame/toaplan/toaplan2.cpp:429]: It’s good practice to call the base `toaplan2_state::machine_start` even if it’s empty at the moment.  If for some reason it needs to be implemented in the future, it’s more convenient if the derived classes don’t all have to be changed at that point.

File: src/mame/toaplan/toaplan2.cpp [hunk 3]
```diff
@@ -431,14 +436,14 @@ WRITE_LINE_MEMBER(toaplan2_state::toaplan2_reset)
...
 
 
-MACHINE_RESET_MEMBER(toaplan2_state,ghox)
+void ghox_state::machine_reset()
 {
 	m_old_p1_paddle_h = 0;
 	m_old_p2_paddle_h = 0;
 }
```
Comment (cuavas) [inline src/mame/toaplan/toaplan2.cpp:443]: Likewise, it’s a good idea to call `toaplan2_state::machine_reset` here.

File: src/mame/toaplan/toaplan2.cpp [hunk 4]
```diff
@@ -661,11 +666,11 @@ READ_LINE_MEMBER(toaplan2_state::c2map_r)
...
-	return m_soundlatch->pending_r() ? 0x00 : 0x01;
+	return m_soundlatch[0]->pending_r() ? 0x00 : 0x01;
 }
 
 
-u16 toaplan2_state::ghox_p1_h_analog_r()
+u16 ghox_state::ghox_p1_h_analog_r()
 {
 	const s8 new_value = m_io_pad[0]->read();
 	if (new_value == m_old_p1_paddle_h) return 0;
```
Comment (cuavas) [inline src/mame/toaplan/toaplan2.cpp:676]: These two functions seem overly complex.  The subtraction will yield zero if the old value and new value are identical anyway so the extra `if` is unnecessary.  You could reduce it to:
```c++
u16 ghox_state::ghox_p1_h_analog_r()
{
	const s8 new_value = m_io_pad[0]->read();
	const s8 result = new_value - m_old_p1_paddle_h;
	if (!machine().side_effects_disabled())
		m_old_p1_paddle_h = new_value;
	return result;
}
```

Also, if you turned the old paddle values into an array, you could use a single function template for both cases and avoid the duplicated code.

**Comment by Tafoid** [comment]
bgaregga - toaplan/toaplan2.cpp                                            3:           Required shared pointer ':tx_gfxram' not found
bgareggabl - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
bgareggabla - toaplan/toaplan2.cpp                                         3:           Required shared pointer ':tx_gfxram' not found
bgareggablj - toaplan/toaplan2.cpp                                         3:           Required shared pointer ':tx_gfxram' not found
bgareggacn - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
bgareggahk - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
bgareggak - toaplan/toaplan2.cpp                                           3:           Required shared pointer ':tx_gfxram' not found
bgaregganv - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
bgareggat - toaplan/toaplan2.cpp                                           3:           Required shared pointer ':tx_gfxram' not found
bgareggat2 - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
bgareggatw - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
fixeightbl - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_gfxram' not found
fixeightbl - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_linescroll' not found
fixeightbl - toaplan/toaplan2.cpp                                          3:           Required shared pointer ':tx_lineselect' not found
kingdmgp - toaplan/toaplan2.cpp                                            3:           Required shared pointer ':tx_gfxram' not found
mahoudai - toaplan/toaplan2.cpp                                            3:           Required shared pointer ':tx_gfxram' not found
shippumd - toaplan/toaplan2.cpp                                            3:           Required shared pointer ':tx_gfxram' not found
sstriker - toaplan/toaplan2.cpp                                            3:           Required shared pointer ':tx_gfxram' not found
sstrikerk - toaplan/toaplan2.cpp                                           3:           Required shared pointer ':tx_gfxram' not found

**Comment by cuavas** [comment]
36e731fc3c95be29458bcf8d6c4de128d6192ffb should fix those errors.  Maybe parts of these drivers should be devicified.

**Pull request merged**